File: src/messaging.py
"""
WhatsApp messaging — Step 4 (final) of the digest pipeline.

Reads data/{city}.json (written by collection.py) and sends the weekly
digest via WhatsApp Cloud API. Runs after the repo commit step so that
a messaging failure never prevents GitHub Pages from being updated.

Run: CITY=brisbane WHATSAPP_TOKEN=... WHATSAPP_PHONE_ID=... WHATSAPP_RECIPIENT=... python src/messaging.py
"""

import json
import os
from datetime import date
from pathlib import Path

import httpx

from common import CATEGORY_EMOJI, TOP_PICK_THRESHOLD, fmt_date


CITY      = os.environ["CITY"]
WA_TOKEN  = os.environ["WHATSAPP_TOKEN"]
WA_PHONE_ID = os.environ["WHATSAPP_PHONE_ID"]
WA_TO     = [n.strip() for n in os.environ["WHATSAPP_RECIPIENT"].split(",") if n.strip()]

MAX_CHARS = 4000  # WhatsApp hard limit is 4096; safe margin


# ---------------------------------------------------------------------------
# Formatting helpers
# ---------------------------------------------------------------------------

def _tags_line(tags: list) -> str:
    return "  🏷 " + " · ".join(str(t) for t in tags[:6]) + "\n"


def _entry_top_pick(e: dict) -> str:
    cat    = e.get("category", "Community / Other")
    emoji  = CATEGORY_EMOJI.get(cat, "📌")
    cost   = e.get("cost", "See link")
    cost_s = "Free ✓" if cost.lower() == "free" else cost
    tags   = e.get("tags", [])
    desc   = e.get("description", "")

    s  = f"• *{e.get('title', 'Untitled')}*\n"
    s += f"  {emoji} {e.get('datetime', '—')}  ·  💰 {cost_s}\n"
    s += f"  📍 {e.get('location', '—')}\n"
    if tags:
        s += _tags_line(tags)
    if desc:
        s += f"  {desc}\n"
    s += f"  🔗 {e.get('link', '')}\n\n"
    return s


def _entry_full(e: dict) -> str:
    cost   = e.get("cost", "See link")
    cost_s = "Free ✓" if cost.lower() == "free" else cost
    tags   = e.get("tags", [])
    desc   = e.get("description", "")

    s  = f"• *{e.get('title', 'Untitled')}*\n"
    s += f"  📆 {e.get('datetime', '—')}  ·  💰 {cost_s}\n"
    s += f"  📍 {e.get('location', '—')}\n"
    if tags:
        s += _tags_line(tags)
    if desc:
        s += f"  {desc}\n"
    s += f"  🔗 {e.get('link', '')}\n\n"
    return s


def _append(messages: list[str], current: str, entry: str, continuation_header: str) -> str:
    if len(current) + len(entry) > MAX_CHARS:
        messages.append(current.strip())
        return continuation_header + entry
    return current + entry


def format_whatsapp(
    events: list[dict], monday: date, sunday: date, city_name: str
) -> list[str]:
    if not events:
        return [
            f"📅 *{city_name} This Week* ({fmt_date(monday)} – {fmt_date(sunday)})\n\n"
            "No events found this week. Check back next Monday!"
        ]

    top_picks = [e for e in events if e.get("score", 0) >= TOP_PICK_THRESHOLD]
    remaining = [e for e in events if e.get("score", 0) < TOP_PICK_THRESHOLD]

    messages: list[str] = []

    header = (
        f"📅 *{city_name} This Week*\n"
        f"{fmt_date(monday)} – {fmt_date(sunday)}\n"
        f"⭐ {len(top_picks)} top picks  ·  {len(events)} events found\n"
        f"{'─' * 28}\n\n"
        f"⭐ *TOP PICKS*\n\n"
    )
    current = header

    for e in top_picks:
        current = _append(messages, current, _entry_top_pick(e), "⭐ *TOP PICKS (cont.)*\n\n")

    if current.strip():
        messages.append(current.strip())

    if not remaining:
        return messages

    by_cat: dict[str, list[dict]] = {}
    for e in remaining:
        by_cat.setdefault(e.get("category", "Community / Other"), []).append(e)

    current = "📋 *All Events This Week*\n\n"

    for cat, cat_events in by_cat.items():
        emoji      = CATEGORY_EMOJI.get(cat, "📌")
        cat_header = f"*{emoji} {cat}*\n"

        if len(current) + len(cat_header) > MAX_CHARS:
            messages.append(current.strip())
            current = cat_header
        else:
            current += cat_header

        for e in cat_events:
            current = _append(messages, current, _entry_full(e), "")

    if current.strip():
        messages.append(current.strip())

    return messages


# ---------------------------------------------------------------------------
# WhatsApp Cloud API
# ---------------------------------------------------------------------------

def send_whatsapp(text: str, recipient: str) -> None:
    print(f"\n── Message preview ──\n{text}\n─────────────────────\n")

    url = f"https://graph.facebook.com/v19.0/{WA_PHONE_ID}/messages"
    payload = {
        "messaging_product": "whatsapp",
        "to": recipient,
        "type": "text",
        "text": {"body": text, "preview_url": False},
    }
    headers = {
        "Authorization": f"Bearer {WA_TOKEN}",
        "Content-Type": "application/json",
    }
    r = httpx.post(url, json=payload, headers=headers, timeout=30)
    print(f"→ WhatsApp API response: {r.status_code} — {r.text[:300]}")
    if r.status_code != 200:
        raise RuntimeError(f"WhatsApp API error {r.status_code}: {r.text}")
    print(f"→ WhatsApp message sent ({len(text)} chars)")


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

def main() -> None:
    json_path = Path(__file__).parent.parent / "data" / f"{CITY}.json"
    if not json_path.exists():
        raise SystemExit(f"✗ {json_path} not found — run collection.py first.")

    payload   = json.loads(json_path.read_text(encoding="utf-8"))
    events    = payload["events"]
    monday    = date.fromisoformat(payload["week_start"])
    sunday    = date.fromisoformat(payload["week_end"])
    city_name = payload["city"]

    print(f"Messaging — {city_name} — {fmt_date(monday)} to {fmt_date(sunday)}")
    print("=" * 50)

    messages = format_whatsapp(events, monday, sunday, city_name)
    print(f"→ Digest split into {len(messages)} WhatsApp message(s)")
    print(f"→ Sending to {len(WA_TO)} recipient(s): {', '.join(WA_TO)}")

    for recipient in WA_TO:
        for i, msg in enumerate(messages, 1):
            print(f"→ [{recipient}] Sending message {i}/{len(messages)}…")
            send_whatsapp(msg, recipient)

    print("✓ Messaging complete.")


if __name__ == "__main__":
    main()

File: src/messaging.ts
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";

import {
	CATEGORY_EMOJI,
	DATA_ROOT,
	fmtDate,
	requireEnv,
	TOP_PICK_THRESHOLD,
} from "./common.ts";

const CITY = requireEnv("CITY");
const WA_TOKEN = requireEnv("WHATSAPP_TOKEN");
const WA_PHONE_ID = requireEnv("WHATSAPP_PHONE_ID");
const WA_TO = requireEnv("WHATSAPP_RECIPIENT")
	.split(",")
	.map((n) => n.trim())
	.filter(Boolean);

const MAX_CHARS = 4000;

type Event = Record<string, unknown>;

function tagsLine(tags: string[]): string {
	return `  🏷 ${tags.slice(0, 6).join(" · ")}\n`;
}

function entryTopPick(e: Event): string {
	const cat = (e.category as string) ?? "Community / Other";
	const emoji = CATEGORY_EMOJI[cat] ?? "📌";
	const cost = (e.cost as string) ?? "See link";
	const costS = cost.toLowerCase() === "free" ? "Free ✓" : cost;
	const tags = (e.tags as string[]) ?? [];
	const desc = (e.description as string) ?? "";

	let s = `• *${(e.title as string) ?? "Untitled"}*\n`;
	s += `  ${emoji} ${(e.datetime as string) ?? "—"}  ·  💰 ${costS}\n`;
	s += `  📍 ${(e.location as string) ?? "—"}\n`;
	if (tags.length > 0) s += tagsLine(tags);
	if (desc) s += `  ${desc}\n`;
	s += `  🔗 ${(e.link as string) ?? ""}\n\n`;
	return s;
}

function entryFull(e: Event): string {
	const cost = (e.cost as string) ?? "See link";
	const costS = cost.toLowerCase() === "free" ? "Free ✓" : cost;
	const tags = (e.tags as string[]) ?? [];
	const desc = (e.description as string) ?? "";

	let s = `• *${(e.title as string) ?? "Untitled"}*\n`;
	s += `  📆 ${(e.datetime as string) ?? "—"}  ·  💰 ${costS}\n`;
	s += `  📍 ${(e.location as string) ?? "—"}\n`;
	if (tags.length > 0) s += tagsLine(tags);
	if (desc) s += `  ${desc}\n`;
	s += `  🔗 ${(e.link as string) ?? ""}\n\n`;
	return s;
}

function append(
	messages: string[],
	current: string,
	entry: string,
	continuationHeader: string,
): string {
	if (current.length + entry.length > MAX_CHARS) {
		messages.push(current.trim());
		return continuationHeader + entry;
	}
	return current + entry;
}

function formatWhatsapp(
	events: Event[],
	monday: Date,
	sunday: Date,
	cityName: string,
): string[] {
	if (events.length === 0) {
		return [
			`📅 *${cityName} This Week* (${fmtDate(monday)} – ${fmtDate(sunday)})\n\n` +
				"No events found this week. Check back next Monday!",
		];
	}

	const topPicks = events.filter(
		(e) => ((e.score as number) ?? 0) >= TOP_PICK_THRESHOLD,
	);
	const remaining = events.filter(
		(e) => ((e.score as number) ?? 0) < TOP_PICK_THRESHOLD,
	);

	const messages: string[] = [];

	const header =
		`📅 *${cityName} This Week*\n` +
		`${fmtDate(monday)} – ${fmtDate(sunday)}\n` +
		`⭐ ${topPicks.length} top picks  ·  ${events.length} events found\n` +
		`${"─".repeat(28)}\n\n` +
		"⭐ *TOP PICKS*\n\n";

	let current = header;
	for (const e of topPicks) {
		current = append(
			messages,
			current,
			entryTopPick(e),
			"⭐ *TOP PICKS (cont.)*\n\n",
		);
	}
	if (current.trim()) messages.push(current.trim());

	if (remaining.length === 0) return messages;

	const byCat: Record<string, Event[]> = {};
	for (const e of remaining) {
		const cat = (e.category as string) ?? "Community / Other";
		if (!byCat[cat]) byCat[cat] = [];
		byCat[cat].push(e);
	}

	current = "📋 *All Events This Week*\n\n";
	for (const [cat, catEvents] of Object.entries(byCat)) {
		const emoji = CATEGORY_EMOJI[cat] ?? "📌";
		const catHeader = `*${emoji} ${cat}*\n`;

		if (current.length + catHeader.length > MAX_CHARS) {
			messages.push(current.trim());
			current = catHeader;
		} else {
			current += catHeader;
		}

		for (const e of catEvents) {
			current = append(messages, current, entryFull(e), "");
		}
	}
	if (current.trim()) messages.push(current.trim());

	return messages;
}

async function sendWhatsapp(text: string, recipient: string): Promise<void> {
	console.log(`\n── Message preview ──\n${text}\n─────────────────────\n`);

	const url = `https://graph.facebook.com/v19.0/${WA_PHONE_ID}/messages`;
	const res = await fetch(url, {
		method: "POST",
		headers: {
			Authorization: `Bearer ${WA_TOKEN}`,
			"Content-Type": "application/json",
		},
		body: JSON.stringify({
			messaging_product: "whatsapp",
			to: recipient,
			type: "text",
			text: { body: text, preview_url: false },
		}),
	});

	const body = await res.text();
	console.log(`→ WhatsApp API response: ${res.status} — ${body.slice(0, 300)}`);
	if (!res.ok) {
		throw new Error(`WhatsApp API error ${res.status}: ${body}`);
	}
	console.log(`→ WhatsApp message sent (${text.length} chars)`);
}

async function main(): Promise<void> {
	const jsonPath = join(DATA_ROOT, `${CITY}.json`);
	if (!existsSync(jsonPath)) {
		throw new Error(`✗ ${jsonPath} not found — run collection.ts first.`);
	}

	const payload = JSON.parse(readFileSync(jsonPath, "utf-8")) as Record<
		string,
		unknown
	>;
	const events = (payload.events as Event[]) ?? [];
	const monday = new Date(payload.week_start as string);
	const sunday = new Date(payload.week_end as string);
	const cityName = payload.city as string;

	console.log(
		`Messaging — ${cityName} — ${fmtDate(monday)} to ${fmtDate(sunday)}`,
	);
	console.log("=".repeat(50));

	const messages = formatWhatsapp(events, monday, sunday, cityName);
	console.log(`→ Digest split into ${messages.length} WhatsApp message(s)`);
	console.log(`→ Sending to ${WA_TO.length} recipient(s): ${WA_TO.join(", ")}`);

	for (const recipient of WA_TO) {
		for (let i = 0; i < messages.length; i++) {
			console.log(
				`→ [${recipient}] Sending message ${i + 1}/${messages.length}…`,
			);
			await sendWhatsapp(messages[i], recipient);
		}
	}

	console.log("✓ Messaging complete.");
}

await main();
